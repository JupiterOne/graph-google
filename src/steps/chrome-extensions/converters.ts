import generateEntityKey from '../../utils/generateEntityKey';
import { entities } from '../../constants';
import {
  Entity,
  RelationshipClass,
  createDirectRelationship,
  createIntegrationEntity,
} from '@jupiterone/integration-sdk-core';
import { chromemanagement_v1 } from 'googleapis';

export function createChromeExtensionEntity(
  data: chromemanagement_v1.Schema$GoogleChromeManagementV1InstalledApp,
) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: generateEntityKey(entities.CHROME_EXTENSION._type, data.appId),
        _type: entities.CHROME_EXTENSION._type,
        _class: entities.CHROME_EXTENSION._class,
        id: data.appId || 'UNKNOWN',
        displayName: data.displayName || data.appId || 'UNKNOWN',
        name: data.displayName,
        description: data.description,
        installType: data.appInstallType,
        source: data.appSource,
        type: data.appType,
        deviceCount: data.browserDeviceCount ?? null,
        disabled: data.disabled,
        uri: data.homepageUri,
        permissions: data.permissions?.join(',') || '',
      },
    },
  });
}

export function createChromeOsDeviceInstalledChromeExtensionRelationship(params: {
  deviceEntity: Entity;
  extensionEntity: Entity;
}) {
  return createDirectRelationship({
    _class: RelationshipClass.INSTALLED,
    from: params.deviceEntity,
    to: params.extensionEntity,
  });
}
