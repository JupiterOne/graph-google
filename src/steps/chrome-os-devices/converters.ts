import generateEntityKey from '../../utils/generateEntityKey';
import { admin_directory_v1 } from 'googleapis';
import { entities } from '../../constants';
import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  parseTimePropertyValue,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';

export function createChromeOSDeviceEntity(
  data: admin_directory_v1.Schema$ChromeOsDevice,
) {
  const deviceName = data.deviceId as string;

  const serialNumber = data.serialNumber;
  const lastSeenOn = parseTimePropertyValue(data.lastSync);

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: generateEntityKey(entities.CHROME_OS_DEVICE._type, deviceName),
        _type: entities.CHROME_OS_DEVICE._type,
        _class: entities.CHROME_OS_DEVICE._class,
        id: deviceName,
        category: 'chrome-os-device',
        make: data.model,
        name: deviceName,
        status: data.status,
        model: data.model,
        serial: data.serialNumber,
        serialNumber,
        version: data.osVersion,
        supportEndDate: data.supportEndDate,
        platformVersion: data.platformVersion,
        firmwareVersion: data.firmwareVersion,
        macAddress: data.macAddress,
        lastSync: lastSeenOn,
        lastSeenOn,
        lastEnrollementTime: parseTimePropertyValue(data.lastEnrollmentTime),
        bootMode: data.bootMode,
        tpmFamily: data.tpmVersionInfo?.family,
        tpmFirmwareVersion: data.tpmVersionInfo?.firmwareVersion,
        tpmManufacturer: data.tpmVersionInfo?.manufacturer,
        tpmSpecLevel: data.tpmVersionInfo?.specLevel,
        tpmModel: data.tpmVersionInfo?.tpmModel,
        tpmVendorId: data.tpmVersionInfo?.vendorSpecific,
      },
    },
  });
}

export function createAccountManagesChromeOSDeviceRelationship(params: {
  accountEntity: Entity;
  deviceEntity: Entity;
}) {
  return createDirectRelationship({
    _class: RelationshipClass.MANAGES,
    from: params.accountEntity,
    to: params.deviceEntity,
  });
}
