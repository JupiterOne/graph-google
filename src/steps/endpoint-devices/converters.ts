import generateEntityKey from '../../utils/generateEntityKey';
import { cloudidentity_v1 } from 'googleapis';
import { entities, relationships } from '../../constants';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
  Entity,
  createDirectRelationship,
} from '@jupiterone/integration-sdk-core';

export function createDeviceEntity(
  data: cloudidentity_v1.Schema$GoogleAppsCloudidentityDevicesV1Device,
) {
  const serialNumber = data.serialNumber || '';
  const lastSeenOn = parseTimePropertyValue(data.lastSyncTime);

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: generateEntityKey(entities.DEVICE._type, data.name),
        _type: entities.DEVICE._type,
        _class: entities.DEVICE._class,
        id: data.name!,
        name: data.name,
        displayName: data.name!,
        createdOn: parseTimePropertyValue(data.createTime),
        lastSyncedOn: lastSeenOn,
        ownerType: data.ownerType,
        model: data.model || '',
        osVersion: data.osVersion,
        deviceType: data.deviceType,
        securityPatchedOn: parseTimePropertyValue(data.securityPatchTime),
        encryptionState: data.encryptionState,
        encrypted: data.encryptionState === 'ENCRYPTED',
        category: 'endpoint',
        make: data.manufacturer || '',
        serial: serialNumber,
        serialNumber,
        // Monkeypatch: deviceId field is not included in type but is included in response.
        // ** deviceId is a required field in J1 Device schema.
        deviceId: (data as any).deviceId || '',
        byod: data.ownerType === 'BYOD',
        assetTag: data.assetTag,
        imei: data.imei,
        meid: data.meid,
        wifiMacAddresses: data.wifiMacAddresses,
        macAddress: data.wifiMacAddresses,
        networkOperator: data.networkOperator,
        releaseVersion: data.releaseVersion,
        brand: data.brand,
        buildNumber: data.buildNumber,
        kernelVersion: data.kernelVersion,
        basebandVersion: data.basebandVersion,
        enabledDeveloperOptions: data.enabledDeveloperOptions,
        otherAccounts: data.otherAccounts,
        enabledUsbDebugging: data.enabledUsbDebugging,
        bootloaderVersion: data.bootloaderVersion,
        managementState: data.managementState,
        compromisedState: data.compromisedState,
        androidEnabledUnknownSources:
          data.androidSpecificAttributes?.enabledUnknownSources,
        androidSupportsWorkProfile:
          data.androidSpecificAttributes?.supportsWorkProfile,
        androidOwnerProfileAccount:
          data.androidSpecificAttributes?.ownerProfileAccount,
        androidOwnershipPrivilege:
          data.androidSpecificAttributes?.ownershipPrivilege,
      },
    },
  });
}

export function createAccountManagesDeviceRelationship(params: {
  accountEntity: Entity;
  deviceEntity: Entity;
}) {
  return createDirectRelationship({
    _class: relationships.ACCOUNT_MANAGES_DEVICE._class,
    from: params.accountEntity,
    to: params.deviceEntity,
  });
}
