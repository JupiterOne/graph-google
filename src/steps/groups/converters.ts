import generateEntityKey from '../../utils/generateEntityKey';
import {
  createIntegrationEntity,
  createDirectRelationship,
  RelationshipClass,
  Entity,
  createMappedRelationship,
  RelationshipDirection,
} from '@jupiterone/integration-sdk-core';
import { admin_directory_v1 } from 'googleapis';
import { entities, relationships } from '../../constants';

export enum MemberType {
  CUSTOMER = 'CUSTOMER',
  EXTERNAL = 'EXTERNAL',
  GROUP = 'GROUP',
  USER = 'USER',
}

export function createGroupEntity(data: admin_directory_v1.Schema$Group) {
  const groupId = data.id as string;

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: generateEntityKey(entities.GROUP._type, groupId),
        _type: entities.GROUP._type,
        _class: entities.GROUP._class,
        id: groupId,
        adminCreated: data.adminCreated,
        directMembersCount: data.directMembersCount,
        displayName: data.name as string,
        email: data.email,
        kind: data.kind,
        name: data.name,
        description: data.description,
      },
    },
  });
}

export function createGroupHasGroupMappedRelationship(
  sourceGroupEntity: Entity,
  groupMember: admin_directory_v1.Schema$Member,
) {
  const targetGroupEntityKey = generateEntityKey(
    'group',
    groupMember.id as string, // Google doesn't ensure the uniqueness of emails in their groups.
  );

  return createMappedRelationship({
    _class: RelationshipClass.HAS,
    _key: `${sourceGroupEntity._key}_has_${targetGroupEntityKey}`,
    _type: relationships.GROUP_HAS_USER._type,
    _mapping: {
      relationshipDirection: RelationshipDirection.FORWARD,
      sourceEntityKey: sourceGroupEntity._key,
      skipTargetCreation: false,
      targetFilterKeys: [['_type', 'email']],
      targetEntity: {
        _type: entities.GROUP._type,
        email: groupMember.email,
      },
    },
  });
}

export function createGroupHasGroupRelationship(params: {
  sourceGroupEntity: Entity;
  targetGroupEntity: Entity;
  groupMember: admin_directory_v1.Schema$Member;
}) {
  const { sourceGroupEntity, targetGroupEntity, groupMember } = params;

  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: sourceGroupEntity,
    to: targetGroupEntity,
    properties: {
      deliverySettings: groupMember.delivery_settings,
      email: groupMember.email,
      id: groupMember.id,
      kind: groupMember.kind,
      role: groupMember.role,
      status: groupMember.status,
      type: groupMember.type,
    },
  });
}

// This means that there is no user found in this integration with user groupMember.id -- ie: this is an EXTERNAL group member
export function createGroupHasUserMappedRelationship(
  sourceGroupEntity: Entity,
  groupMember: admin_directory_v1.Schema$Member,
) {
  const targetUserEntityKey = generateEntityKey(
    'user',
    groupMember.id as string, // Google doesn't ensure the uniqueness of emails in their groups.
  );

  // Create a mapped relationship
  return createMappedRelationship({
    _class: RelationshipClass.HAS,
    _key: `${sourceGroupEntity._key}_has_${targetUserEntityKey}`,
    _type: relationships.GROUP_HAS_USER._type,
    _mapping: {
      relationshipDirection: RelationshipDirection.FORWARD,
      sourceEntityKey: sourceGroupEntity._key,
      skipTargetCreation: false,
      targetFilterKeys: [['_type', 'email']],
      targetEntity: {
        _type: entities.USER._type,
        email: groupMember.email,
      },
    },
  });
}

export function createGroupHasUserRelationship(params: {
  sourceGroupEntity: Entity;
  targetUserEntity: Entity;
  groupMember: admin_directory_v1.Schema$Member;
}) {
  const { sourceGroupEntity, targetUserEntity, groupMember } = params;

  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: sourceGroupEntity,
    to: targetUserEntity,
    properties: {
      deliverySettings: groupMember.delivery_settings,
      email: groupMember.email,
      id: groupMember.id,
      kind: groupMember.kind,
      role: groupMember.role,
      status: groupMember.status,
      type: groupMember.type,
    },
  });
}

export function createAccountHasGroupRelationship(params: {
  accountEntity: Entity;
  groupEntity: Entity;
}) {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: params.accountEntity,
    to: params.groupEntity,
  });
}
