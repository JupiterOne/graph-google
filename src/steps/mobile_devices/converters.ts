import generateEntityKey from '../../utils/generateEntityKey';
import { admin_directory_v1 } from 'googleapis';
import { entities } from '../../constants';
import {
  createIntegrationEntity,
  parseTimePropertyValue,
  Entity,
  RelationshipClass,
  createDirectRelationship,
} from '@jupiterone/integration-sdk-core';

function firstOrUndefined(v: any | any[]) {
  return Array.isArray(v) ? v[0] : undefined;
}

export function createMobileDeviceEntity(
  data: admin_directory_v1.Schema$MobileDevice,
) {
  const deviceName = data.deviceId as string;
  const name = firstOrUndefined(data.name);

  const serialNumber = data.serialNumber;
  const lastSeenOn = parseTimePropertyValue(data.lastSync);

  return createIntegrationEntity({
    entityData: {
      // The raw data on mobile devices can be very large, so we've removed it.
      source: {},
      assign: {
        _key: generateEntityKey(entities.MOBILE_DEVICE._type, deviceName),
        _type: entities.MOBILE_DEVICE._type,
        _class: entities.MOBILE_DEVICE._class,
        id: deviceName,
        name:
          name ||
          `${data.email?.[0] || 'Unknown User'}'s ${data.model || 'Device'}`,
        displayName:
          name ||
          `${data.email?.[0] || 'Unknown User'}'s ${data.model || 'Device'}`,
        category: 'mobile',
        make: data.model,
        model: data.model,
        serial: serialNumber,
        serialNumber,
        adbStatus: data.adbStatus,
        brand: data.brand,
        deviceCompromisedStatus: data.deviceCompromisedStatus,
        deviceId: data.deviceId,
        email: firstOrUndefined(data.email),
        encryptionStatus: data.encryptionStatus,
        firstSyncOn: parseTimePropertyValue(data.firstSync),
        hardware: data.hardware,
        hardwareId: data.hardwareId,
        imei: data.imei,
        lastSyncOn: lastSeenOn,
        lastSeenOn,
        manufacturer: data.manufacturer,
        ownerName: name,
        os: data.os,
        status: data.status,
        type: data.type,
        wifiMacAddress: data.wifiMacAddress,
        macAddress: data.wifiMacAddress,
        userAgent: data.userAgent,
      },
    },
  });
}

export function createAccountManagesMobileDeviceRelationship(params: {
  accountEntity: Entity;
  deviceEntity: Entity;
}) {
  return createDirectRelationship({
    _class: RelationshipClass.MANAGES,
    from: params.accountEntity,
    to: params.deviceEntity,
  });
}
