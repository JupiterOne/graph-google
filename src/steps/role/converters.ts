import generateEntityKey from '../../utils/generateEntityKey';
import { entities } from '../../constants';
import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';

// TODO/Rick: Interface is still for Account, update to role
interface CreateRoleEntityParams {
  account: { googleAccountId: string; name: string };
  domainNames: string[];
  primaryDomain?: string;
}

export function createRoleEntity(data: CreateRoleEntityParams) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: entities.ROLE._class,
        _key: generateEntityKey(
          entities.ROLE._type,
          data.account.googleAccountId,
        ),
        _type: entities.ROLE._type,
        // TODO/Rick: Shape
        displayName: data.account.name,
        name: data.account.name,
        domains: data.domainNames,
        primaryDomain: data.primaryDomain,
        accountId: data.primaryDomain,
        id: data.account.googleAccountId,
        vendor: 'Google',
      },
    },
  });
}
