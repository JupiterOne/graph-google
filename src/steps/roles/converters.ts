import generateEntityKey from '../../utils/generateEntityKey';
import { entities } from '../../constants';
import { createIntegrationEntity } from '@jupiterone/integration-sdk-core';
import { admin_directory_v1 } from 'googleapis';
import Schema$Role = admin_directory_v1.Schema$Role;

export function createRoleEntity(role: Schema$Role) {
  return createIntegrationEntity({
    entityData: {
      source: role,
      assign: {
        _class: entities.ROLE._class,
        _type: entities.ROLE._type,
        _key: generateEntityKey(entities.ROLE._type, role.roleId as string),
        // TODO/Rick: Is this the shape we want? I am including more properties for discussion but can whittle this down
        id: role.roleId as string,
        name: role.roleName,
        displayName: role.roleName as string,
        roleName: role.roleName,
        roleDescription: role.roleDescription,
        isSystemRole: role.isSystemRole,
        isSuperAdminRole: role.isSuperAdminRole,
        kind: role.kind,
        vendor: 'Google',
      },
    },
  });
}
