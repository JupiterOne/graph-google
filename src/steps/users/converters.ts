import generateEntityKey, { UnsafeIdKey } from '../../utils/generateEntityKey';
import {
  parseTimePropertyValue,
  createIntegrationEntity,
  Entity,
  createDirectRelationship,
  RelationshipClass,
  convertProperties,
} from '@jupiterone/integration-sdk-core';
import { admin_directory_v1 } from 'googleapis';
import { entities } from '../../constants';

interface GSuiteDataCollection {
  type: string;
}

interface GetCollectionAsFlattendFieldsParams<T extends GSuiteDataCollection> {
  collection: T[];
  suffix: string;
  valueMethod: string;
}

export function getCollectionAsFlattendFields<T extends GSuiteDataCollection>({
  collection,
  suffix,
  valueMethod,
}: GetCollectionAsFlattendFieldsParams<T>): Record<string, any> {
  const flattendRecordFields: Record<string, any> = {};

  collection.forEach((item) => {
    flattendRecordFields[`${item.type}${suffix}`] = item[valueMethod];
    if (/.+@.+\..+/.exec(item[valueMethod])) {
      flattendRecordFields[`${item.type}Email`] = item[valueMethod];
    }
  });

  return flattendRecordFields;
}

export function getUserEntityKey(userId: UnsafeIdKey) {
  return generateEntityKey(entities.USER._type, userId);
}

export function createUserEntity(data: admin_directory_v1.Schema$User) {
  const userId = data.id as string;
  const name = data.name?.fullName || data.primaryEmail;

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _key: getUserEntityKey(userId),
        _type: entities.USER._type,
        _class: entities.USER._class,
        id: userId,
        email: data.primaryEmail,
        emailDomain: getDomain(data),
        name,
        displayName: name,
        username: data.primaryEmail,
        firstName: data.name?.givenName,
        lastName: data.name?.familyName,
        mfaEnabled: !!data.isEnrolledIn2Sv,
        suspended: data.suspended,
        archived: data.archived,
        active: !data.suspended && !data.archived && data.agreedToTerms,
        agreedToTerms: data.agreedToTerms,
        changePasswordAtNextLogin: data.changePasswordAtNextLogin,
        createdOn: parseTimePropertyValue(data.creationTime),
        creationTime: parseTimePropertyValue(data.creationTime),
        deletedOn: parseTimePropertyValue(data.deletionTime),
        deletionTime: parseTimePropertyValue(data.deletionTime),
        lastLoginOn: parseTimePropertyValue(data.lastLoginTime),
        lastLoginTime: parseTimePropertyValue(data.lastLoginTime),
        customerId: data.customerId,
        hashFunction: data.hashFunction,
        includeInGlobalAddressList: data.includeInGlobalAddressList,
        ipWhitelisted: data.ipWhitelisted,
        admin: data.isAdmin || data.isDelegatedAdmin,
        isAdmin: data.isAdmin,
        isDelegatedAdmin: data.isDelegatedAdmin,
        isEnforcedIn2Sv: data.isEnforcedIn2Sv,
        isEnrolledIn2Sv: data.isEnrolledIn2Sv,
        isMailboxSetup: data.isMailboxSetup,
        kind: data.kind,
        orgUnitPath: data.orgUnitPath,
        primaryEmail: data.primaryEmail,
        recoveryEmail: data.recoveryEmail,
        recoveryPhone: data.recoveryPhone,
        suspensionReason: data.suspensionReason,
        thumbnailPhotoUrl: data.thumbnailPhotoUrl,
        aliases: data.aliases,
        // Copy the entire stringified value of `customSchemas` onto the entity,
        // so that it can be queried using a `contains` filter.
        customSchemas: data.customSchemas && JSON.stringify(data.customSchemas),
        githubUsername:
          data.customSchemas &&
          data.customSchemas['Github'] &&
          data.customSchemas['Github']['githubUsername'],
        ...getAddresses(data),
        ...getPhones(data),
        ...getRelations(data),
        ...getExternalIds(data),
        // TODO: This will filter out emails that do not have a "type" property.
        // Should we also add an array of emails to the entity?
        ...getEmails(data),
        ...getManagementInfo(data),
        ...getEmployeeInfo(data),
      },
    },
  });
}

export function createSiteEntity(
  userId: string,
  data: admin_directory_v1.Schema$UserLocation,
) {
  const name = [data.buildingId, data.floorName, data.floorSection].join(', ');

  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _type: entities.SITE._type,
        _class: entities.SITE._class,
        _key: generateEntityKey(
          entities.SITE._type,
          `${userId}_${data.buildingId}`,
        ),
        name,
        displayName: name,
        id: data.buildingId as string,
        type: data.type,
        area: data.area,
        buildingId: data.buildingId,
        floorName: data.floorName,
        floorSection: data.floorSection,
      },
    },
  });
}

export function createSiteHostsUserRelationship(params: {
  siteEntity: Entity;
  userEntity: Entity;
}) {
  return createDirectRelationship({
    // TODO: Change to HOSTS
    _class: RelationshipClass.HAS,
    from: params.siteEntity,
    to: params.userEntity,
  });
}

export function createAccountHasUserRelationship(params: {
  accountEntity: Entity;
  userEntity: Entity;
}) {
  return createDirectRelationship({
    _class: RelationshipClass.HAS,
    from: params.accountEntity,
    to: params.userEntity,
  });
}

function last<T>(arr: T[] | undefined): T | undefined {
  if (!arr || !arr.length) return undefined;
  return arr[arr.length - 1];
}

function getDomain(data: admin_directory_v1.Schema$User): string[] {
  const emailDomain = last(data.primaryEmail?.split('@'));
  if (emailDomain) {
    return [emailDomain];
  }
  return [];
}

function getAddresses(data: admin_directory_v1.Schema$User) {
  return (
    data.addresses &&
    getCollectionAsFlattendFields({
      collection: data.addresses,
      suffix: 'Address',
      valueMethod: 'formatted',
    })
  );
}

function getPhones(data: admin_directory_v1.Schema$User) {
  return (
    data.phones &&
    getCollectionAsFlattendFields({
      collection: data.phones,
      suffix: 'Phone',
      valueMethod: 'value',
    })
  );
}

function getRelations(data: admin_directory_v1.Schema$User) {
  return (
    data.relations &&
    getCollectionAsFlattendFields({
      collection: data.relations,
      suffix: 'Relation',
      valueMethod: 'value',
    })
  );
}

function getExternalIds(data: admin_directory_v1.Schema$User) {
  return (
    data.externalIds &&
    getCollectionAsFlattendFields({
      collection: data.externalIds,
      suffix: 'ExternalId',
      valueMethod: 'value',
    })
  );
}

function getEmails(data: admin_directory_v1.Schema$User) {
  return (
    data.emails &&
    getCollectionAsFlattendFields({
      collection: data.emails.filter((e: { type: string }) => !!e.type),
      suffix: 'Email',
      valueMethod: 'address',
    })
  );
}

interface UserManagementInfo {
  managerEmail?: string;
  manager?: string;
}

function getManagementInfo(
  data: admin_directory_v1.Schema$User,
): UserManagementInfo | undefined {
  const managerRelation = (data as any).managerRelation;

  return (
    managerRelation &&
    (managerRelation.include('@')
      ? { managerEmail: managerRelation }
      : { manager: managerRelation })
  );
}

interface EmployeeInfo {
  title: string;
  customType: string;
  department: string;
  employeeType: string;
  costCenter: string;
}

function getEmployeeInfo(
  data: admin_directory_v1.Schema$User,
): EmployeeInfo | undefined {
  if (!data.organizations || data.organizations.length === 0) {
    return;
  }

  const primaryOrganization = data.organizations.find((o) => !!o.primary);
  const org = primaryOrganization || data.organizations[0];

  const employeeInfo = {
    ...convertProperties(org),
    // `employeeInfo.description` corresponds to "Type of Employee" in the
    // Google Admin UI under "Employee Information" section.
    employeeType: org.description,
  };

  delete employeeInfo.primary;
  return employeeInfo;
}
