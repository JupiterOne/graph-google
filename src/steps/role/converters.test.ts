import { createRoleEntity } from './converters';

describe.skip('#createRoleEntity', () => {
  test('should convert to entity', () => {
    expect(
      createRoleEntity({
        account: {
          googleAccountId: 'abc123',
          name: 'mygoogleyo',
        },
        domainNames: ['jupiterone.com', 'jupiterone.io'],
        primaryDomain: 'jupiterone.com',
      }),
    ).toMatchSnapshot();
  });
});
