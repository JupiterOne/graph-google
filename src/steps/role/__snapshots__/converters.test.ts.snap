// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createRoleEntity should convert to entity 1`] = `
Object {
  "_class": Array [
    "AccessRole",
  ],
  "_key": "google_role_abc123",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "account": Object {
          "googleAccountId": "abc123",
          "name": "mygoogleyo",
        },
        "domainNames": Array [
          "jupiterone.com",
          "jupiterone.io",
        ],
        "primaryDomain": "jupiterone.com",
      },
    },
  ],
  "_type": "google_role",
  "accountId": "jupiterone.com",
  "createdOn": undefined,
  "displayName": "mygoogleyo",
  "domains": Array [
    "jupiterone.com",
    "jupiterone.io",
  ],
  "id": "abc123",
  "name": "mygoogleyo",
  "primaryDomain": "jupiterone.com",
  "vendor": "Google",
}
`;
